tools/undelete-blocks/main.go

```diff
@@ -0,0 +1,471 @@
+	return nil
+}
+
+type tenantFilter func(string) bool
+
+func newTenantFilter(cfg config) tenantFilter {
+	includeTenants := map[string]struct{}{}
+	for _, u := range cfg.includeTenants {
+		includeTenants[u] = struct{}{}
+	}
+	excludeTenants := map[string]struct{}{}
+	for _, u := range cfg.excludeTenants {
+		excludeTenants[u] = struct{}{}
+	}
+
+	return func(tenantID string) bool {
+		if len(includeTenants) > 0 {
+			if _, ok := includeTenants[tenantID]; !ok {
+				return false
+			}
+		}
+		_, ok := excludeTenants[tenantID]
+		return !ok
+	}
+}
+
+func getBlocks(ctx context.Context, cfg config, bucket objtools.Bucket) (map[string][]ulid.ULID, error) {
+	tenantFilter := newTenantFilter(cfg)
+	if cfg.inputFile != "" {
+		switch strings.ToLower(cfg.inputFileFormat) {
+		case "json":
+			return getBlocksFromJSONFile(cfg.inputFile, tenantFilter)
+		case "lines":
+			return getBlocksFromLinesFile(cfg.inputFile, tenantFilter)
+		default:
+			return nil, errors.Errorf("unrecognized input file format: %s", cfg.inputFileFormat)
+		}
+	}
+	return getBlocksFromListing(ctx, bucket, tenantFilter)
+}
+
+// getBlocksFromJSONFile reads a JSON tenant to blockIDs map from the specified file
+func getBlocksFromJSONFile(filePath string, filter tenantFilter) (map[string][]ulid.ULID, error) {
+	f, err := os.Open(filePath)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+
+	b, err := io.ReadAll(f)
+	if err != nil {
+		return nil, err
+	}
+
+	m := make(map[string][]string)
+	if err = json.Unmarshal(b, &m); err != nil {
+		return nil, err
+	}
+
+	for tenant := range m {
+		if !filter(tenant) {
+			delete(m, tenant)
+		}
+	}
+
+	m2 := make(map[string][]ulid.ULID)
+	for k, v := range m {
+		v2 := make([]ulid.ULID, 0, len(v))
+		for _, s := range v {
+			u, err := ulid.Parse(s)
+			if err != nil {
+				return nil, errors.Wrapf(err, "failed to parse a string=%s as a ULID for tenant=%s", s, k)
+			}
+			v2 = append(v2, u)
+		}
+		m2[k] = v2
+	}
+
+	return m2, nil
+}
+
+// getBlocksFromLinesFile reads a file with each line having a tenant and a blockID separated by a space
+func getBlocksFromLinesFile(filePath string, filter tenantFilter) (map[string][]ulid.ULID, error) {
+	f, err := os.Open(filePath)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+
+	m := make(map[string][]ulid.ULID)
+	scanner := bufio.NewScanner(f)
+	for scanner.Scan() {
+		line := scanner.Text()
+		tenant, blockString, found := strings.Cut(line, " ")
+		if !found {
+			return nil, errors.Errorf("no space separating tenant and block in line formatted file: %s", line)
+		}
+		if !filter(tenant) {
+			continue
+		}
+		u, err := ulid.Parse(blockString)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to parse a string=%s as a ULID for tenant=%s", blockString, tenant)
+		}
+		m[tenant] = append(m[tenant], u)
+	}
+	return m, nil
+}
+
+// getBlocksFromListing does a prefixed versioned listing for to find all block prefixes in each unfiltered tenant
```

Comment:
I am really uncomfortable by letting the tool to list tenants and doing the undelete based on filters. Let's move listing tenants and blocks to separate tool. Then we can inspect the output of listing before deciding how to proceed.

Reply:
I removed the listing behavior as the default and forced the user to provide `--blocks-from listing` as an attempt to address your safety concerns. We've talked about adding support for `listblocks` for more precise filtering.